CMakeLists.txt

```diff
@@ -38,150 +36,170 @@ if(NOT FREERTOS_PORT)
+        " PARADIGM_TERN_EE_LARGE           - Compiler: Paradigm      Target: Tern EE large\n"
+        " PARADIGM_TERN_EE_SMALL           - Compiler: Paradigm      Target: Tern EE small\n"
+        " RENESAS_RX100                    - Compiler: Renesas       Target: RX100\n"
+        " RENESAS_RX200                    - Compiler: Renesas       Target: RX200\n"
+        " RENESAS_RX600                    - Compiler: Renesas       Target: RX600\n"
+        " RENESAS_RX600_V2                 - Compiler: Renesas       Target: RX600 v2\n"
+        " RENESAS_RX700_V3_DPFPU           - Compiler: Renesas       Target: RX700 v3 with DPFPU\n"
+        " RENESAS_SH2A_FPU                 - Compiler: Renesas       Target: SH2A with FPU\n"
+        " ROWLEY_MSP430F449                - Compiler: Rowley        Target: MSP430F449\n"
+        " RVDS_ARM_CA9                     - Compiler: RVDS          Target: ARM Cortex-A9\n"
+        " RVDS_ARM_CM0                     - Compiler: RVDS          Target: ARM Cortex-M0\n"
+        " RVDS_ARM_CM3                     - Compiler: RVDS          Target: ARM Cortex-M3\n"
+        " RVDS_ARM_CM4_MPU                 - Compiler: RVDS          Target: ARM Cortex-M4 with MPU\n"
+        " RVDS_ARM_CM4F                    - Compiler: RVDS          Target: ARM Cortex-M4 with FPU\n"
+        " RVDS_ARM_CM7                     - Compiler: RVDS          Target: ARM Cortex-M7\n"
+        " RVDS_ARM7_LPC21XX                - Compiler: RVDS          Target: ARM7 LPC21xx\n"
+        " SDCC_CYGNAL                      - Compiler: SDCC          Target: Cygnal\n"
+        " SOFTUNE_MB91460                  - Compiler: Softune       Target: MB91460\n"
+        " SOFTUNE_MB96340                  - Compiler: Softune       Target: MB96340\n"
+        " TASKING_ARM_CM4F                 - Compiler: Tasking       Target: ARM Cortex-M4 with FPU\n"
+        " CDK_THEAD_CK802                  - Compiler: CDK           Target: T-head CK802\n"
+        " XCC_XTENSA                       - Compiler: XCC           Target: Xtensa\n"
+        " WIZC_PIC18                       - Compiler: WizC          Target: PIC18")
+elseif((FREERTOS_PORT STREQUAL "A_CUSTOM_PORT") AND (NOT TARGET freertos_kernel_port) )
+    message(FATAL_ERROR " FREERTOS_PORT is set to A_CUSTOM_PORT. Please specify the custom port target with all necessary files. For example:\n"
+    " Assuming a directory of:\n"
+    "  FreeRTOSCustomPort/\n"
+    "    CMakeLists.txt\n"
+    "    port.c\n"
+    "    portmacro.h\n"
+    " Where FreeRTOSCustomPort/CMakeLists.txt is a modified version of:\n"
+    "   add_library(freertos_kernel_port STATIC)\n"
+    "   target_sources(freertos_kernel_port\n"
+    "     PRIVATE\n"
+    "       port.c\n"
+    "       portmacro.h)\n"
+    "   target_include_directories(freertos_kernel_port\n"
+    "     PUBLIC\n"
+    "      .)\n"
+    "   taget_link_libraries(freertos_kernel_port\n"
```

Comment:
I am not Cmake expert but is it necessary to link `freertos_kernel` target when `freertos_kernel` links ``freertos_kernel_port` below.

Reply:
It should since port uses kernel headers.

Reply:
I don't think it makes sense to export both the port and kernel as separate STATIC lib targets. 

They're fully intertwined by the preprocessor, so it really only makes sense to export a static lib target for the combination of a config, port, and the kernel. 

If we're going to export targets for each piece separately, they should be INTERFACE lib targets rather than STATIC lib targets.

Reply:
First question: Yes it is necessary to link `freertos_kernel` target to `freertos_kernel_port` as `freertos_kernel_port` consumes the headers in `freertos_kernel`.  There is no issue with circular dependencies since it is linked as  `PRIVATE`.

Second question: Not quite sure what you refer to as exported.  It is necessary to separate the static libs since the `freertos_kernel_port` is select-able and it makes it easier if these are separate.  I was debating adding a CMakeLists.txt per `FREERTOS_PORT` type instead of having it monolithic in the port directory - since this limits the scope of a change to a particular port when a single port changes.

As for exporting each piece separately.  No - you can consume `freertos_kernel` in your `target_link_libraries` at the project level and you will also get both the `freertos_config` and the `freertos_kernel_port` linked too due to the PUBLIC link of those here: https://github.com/phelter/FreeRTOS-Kernel/blob/1822d8eb9aeb5aedb66cbf56e4a7fca58d898009/CMakeLists.txt#L241

A bit of a tutorial on `INTERFACE` lib targets - they `should` only be used for header only targets.  You could separate the `freertos_kernel` into an `INTERFACE` and `STATIC` lib but unless there is a really good reason - i.e. circular dependencies on targets, it is much easier to maintain as a single `target`

Reply:
That is an unconventional way of using an INTERFACE library in CMake, but will still work.

The only thing that you'd need to change in this PR is:

From: `target_sources(freertos_kernel PRIVATE`
To: `target_sources(freertos_kernel PUBLIC`

All other changes are superficial and you do not require an INTERFACE type of target for this purpose.

What this does is it tells the builder to `copy a version of the code into the build directory of whatever will link to freertos_kernel and then compile out of the `build` directory of the target that linked to `freertos_kernel`.  This has many drawbacks.  The first being that you may want multiple other targets consuming `freertos_kernel` in the same application.  In this case `freertos_kernel` will be compiled multiple times for each target that links to it (and has the potential to be compiled differently with different compile options (not something you typically want) for two targets that then get consumed into a single application.  I HIGHLY RECOMMEND that you DO NOT take this approach.

If you would like to generate multiple binaries with different FreeRTOSConfig.h in the same cmake project then this is typically done with multiple compile options, eg:

```bash
cmake -GNinja -DSELECT_FREERTOS_APP=APP1 -B build_app1 .
cmake -GNinja -DSELECT_FREERTOS_APP=APP2 -B build_app2 .
```
In this case the application layer decides which version of `FreeRTOSConfig.h` is used for each APP1 and APP2 in its own `CMakeLists.txt` and the entire directory build_app1 is build with one implementation and the build_app2 directory could be a different implementation.  They both would consume the same code, just in different directories.  This is more common way of providing the feature of multiple binaries with different `FreeRTOSConfig.h` files and would be the suggested way to explain to integrators/application developers.

You could do it the other way as shown in your example of the `RPI Pico SDK` but then for 90% of the use cases where it is only one port that they require, you would be potentially burdening that single compilation with multiple additional compiles, and there is a larger chance an application developer could shoot themselves in the foot with different FreeRTOSConfig.h files for the same application.

This also leads down the path of then supporting a single cmake project wanting two FreeRTOS-Kernels with two different ports - eg. Little-Big ARM based SoC's.  That situation can be handled with the 'multiple compile options` method mentioned above too.

Reply:
Multiple build directories are sometimes a reasonable option, but not the only option. I don't think it makes sense to enforce this on users, especially when it will further complicate adding CI checks for each port and their multitude of different configurations.

If we're set on exporting a freertos_kernel static library target, we should just include the selected port port with it rather than creating two separate targets and implying that the two are not circularly dependent on each other.

Reply:
@paulbartell - If FreeRTOS is wanting to support multiple Ports, then shouldn't the CI checks confirm each port can be compiled - with some minimal checks - eg: `Default, AllEnabled, AllDisabled` - for the portConfig info?  And when a user decides to have their own special `config` and a bug related to the kernel is found in their `freertos_config` configuration - it is an easy fix to confirm the bug by providing that user's config.h files to re-create.

At this moment nothing is enforced on the user/consumer -  The changes here are following the original CMake that was provided.  The only difference is that the user now specify their one and only one `freertos_config` target instead of defining one and only one `FREERTOS_CONFIG_FILE_DIRECTORY` that was specified. One cannot have multiple values of a CMAKE variable for the same build directory.

As mentioned - I'm not sure what you are referring to as `exporting` as this is expected to be compiled with the user's code (as based on the `freertos_config`) so there is no notion of `exporting` - which is typically a common lib that is provided compiled without configuration.  As mentioned above, the `freertos_kernel` target is an aggregate of the `freertos_kernel_port` (the one selected), the `freertos_config` - provided by the user, and the `freertos_kernel` code.  - This has not changed from the original definition of CMake files as they already had an aggregate dependency as mentioned.

Within CMake there is no means of hiding a target from the user, so the `freertos_kernel_port` being defined independently will be visible to the user, and because of the dependencies will compile both `freertos_kernel` and `freertos_kernel_port`.

If you wanted to remove what you are referring to as `circular-dependencies` one could have:
- create a `freertos_kernel_port_api` which would only be the header files that are consumed by the `freertos_kernel_port_x`
- include `freertos_kernel_port_api` as part of both `freertos_kernel_port_x` and `freertos_kernel`
- Deal with the confusion of having yet another target.
Typically the `api` is there if you intend on supporting more than one implementation - you define `api` so that that is the base for any dependencies.

But this is all out of the scope of what I was trying to achieve.

1. Support for a target instead of a single PATH to the config files.
2. Support for a specialized port that isn't defined as one of the defaults in the FreeRTOS-Kernel
- Where the second one would mean writing your own CMake if you wanted to do so instead of relying on what is already there.

Reply:
@phelter : I understand your position and I'm very glad to have you contributing. The CMakeLists.txt is not perfect, but we're making good progress.

One reuqest, Could you make the freertos_config target optional and link to it if and only if it exists? With that change, I can approve this PR.

Reply:
@paulbartell - Those changes are now in, and I also identified the deprecated `FREERTOS_CONFIG_FILE_DIRECTORY` portions of the cmake, and removed the auto generation of the `freertos_config` target if a user had `FREERTOS_CONFIG_FILE_DIRECTORY` defined prior but no `freertos_config`.